bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/HydrawiseCloudHandler.java

```diff
@@ -0,0 +1,246 @@
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.openhab.binding.hydrawise.internal.api.HydrawiseAuthenticationException;
+import org.openhab.binding.hydrawise.internal.api.HydrawiseCloudApiClient;
+import org.openhab.binding.hydrawise.internal.api.HydrawiseCommandException;
+import org.openhab.binding.hydrawise.internal.api.HydrawiseConnectionException;
+import org.openhab.binding.hydrawise.internal.api.model.Controller;
+import org.openhab.binding.hydrawise.internal.api.model.CustomerDetailsResponse;
+import org.openhab.binding.hydrawise.internal.api.model.Forecast;
+import org.openhab.binding.hydrawise.internal.api.model.Relay;
+import org.openhab.binding.hydrawise.internal.api.model.StatusScheduleResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link HydrawiseCloudHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Dan Cunningham - Initial contribution
+ */
+@NonNullByDefault
+public class HydrawiseCloudHandler extends HydrawiseHandler {
+    private final Logger logger = LoggerFactory.getLogger(HydrawiseCloudHandler.class);
+    private HydrawiseCloudApiClient client;
+    int controllerId;
+    /**
+     * 74.2 F
+     */
+    private static final Pattern TEMPERATURE_PATTERN = Pattern.compile("^(\\d{1,3}.?\\d?)\\s([C,F])");
+    /**
+     * 9 mph
+     */
+    private static final Pattern WIND_SPEED_PATTERN = Pattern.compile("^(\\d{1,3})\\s([a-z]{3})");
+
+    public HydrawiseCloudHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.client = new HydrawiseCloudApiClient(httpClient);
+    }
+
+    @Override
+    protected void configure()
+            throws NotConfiguredException, HydrawiseConnectionException, HydrawiseAuthenticationException {
+        HydrawiseCloudConfiguration configuration = getConfig().as(HydrawiseCloudConfiguration.class);
+
+        this.refresh = Math.max(configuration.refresh, MIN_REFRESH_SECONDS);
+
+        client.setApiKey(configuration.apiKey);
+
+        CustomerDetailsResponse customerDetails = client.getCustomerDetails();
+
+        List<Controller> controllers = customerDetails.getControllers();
+        if (controllers.size() == 0) {
+            throw new NotConfiguredException("No controllers found on account");
+        }
+
+        Controller controller = null;
+        // try and use ID from user configuration
+        if (configuration.controllerId != null) {
+            controller = getController(configuration.controllerId.intValue(), controllers);
+            if (controller == null) {
+                throw new NotConfiguredException("No controller found for id " + configuration.controllerId);
+            }
+        } else {
+            // try and use ID from saved property
+            if (StringUtils.isNotBlank(getThing().getProperties().get(PROPERTY_CONTROLLER_ID))) {
```

Comment:
Please assign `getThing().getProperties().get(PROPERTY_CONTROLLER_ID)` to some variable instead of fetching is 3 times. It might be not in-efficient now but who knows what changes.